### PR TITLE
improve DetailsList row performance

### LINKED ISSUE #3853: DetailsList: Very slow performance when there are multiple instances.
### Bug Report

- __Package version(s)__: latest
- __Browser and OS versions__: Windows, chrome, edge, IE11

#### Priorities and help requested (not applicable if asking question):

Are you willing to submit a PR to fix? No

Requested priority: High

#### Describe the issue:

On the [DetailsList example page](https://developer.microsoft.com/en-us/fabric#/components/detailslist) or where there are many instances of DetailsList, the performance is taking a huge hit. This may be relevant https://github.com/OfficeDev/office-ui-fabric-react/issues/3790#issuecomment-360213990

#### Actual behavior:

Page slows down, clicks are delayed, resizing takes forever, etc.

#### Expected behavior:

Better performance

### PR DESCRIPTION
**Pull request by OneCyrus**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #3853 
- [x] Include a change request file using `$ npm run change`

#### Description of changes

added shouldComponentUpdate to detailsrow to only render the row when the props change

#### Focus areas to test

any edge cases where rows should render again when not changing the props

### CODE REVIEW

**Comment by JasonGore** [comment]
I gave this change a try out of curiosity and I'm not seeing any noticeable differences in behavior. Can you quantify the improvement in any way?

**Comment by dzearing** [comment]
@JasonGore This change may reduce the number of renders that occur on scrolls. @OneCyrus it would be good to enumerate exactly the steps you were testing with.

You can probably see it by providing a cell render that renders the "current render count"... today it may redundantly render things at certain times, It's possible this could avoid some re-renders if the parent re-renders and passes the same content into the List.

This particular change runs the risk of potentially avoiding an update that someone was previously relying on. I'm wondering if this should be opt in to avoid a potential break.

**Comment by JasonGore** [comment]
@dzearing I think it should be up to the author to quantify the effects seen and/or improvement gain. 😄 

If stores don't mutate state (as is standard in React), is it not enough just to do a reference comparison in cases like these?

**Comment by dzearing** [comment]
The screener tests reveal that this breaks selection. We would need to address this to consider this PR.

**Comment by cliffkoh** [comment]
So, this means that even within the realm of the component, it doesn't treat the props as immutable. It's very likely that this little change might end up breaking partners. We therefore would need to make this an opt-in optimization to minimize disruption... (in addition to fixing the selection regression caught by screener tests)

**Comment by aditima** [comment]
@OneCyrus, as Cliff and David mentioned above we'd require the following changes to consider your PR:
1. Fix selection as pointer out by screener breaks
2. Add this optimization as an opt-in as this relies on props being immutable, and might end up breaking partners who don't follow that assumption

**Comment by OneCyrus** [comment]
I fixed the selection issue and added a feature flag. the main issue with the details list is the architecture of the data flow. currently the details row needs two rendering for changing the selection state. a proper refactoring could help this component quite a bit. but with this change there should be a big improvement already.

**Comment by dzearing** [comment]
@OneCyrus I agree it absolutely needs a refactoring. I've started looking at the underlying List component first, as that will have the biggest impact on perf improvement.

**Comment by dzearing** [comment]
@aditima Are you ok with this change?

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts [hunk 1]
```diff
@@ -245,6 +245,11 @@ export interface IDetailsListProps extends React.Props<DetailsList>, IWithViewpo
    * Whether or not the selection zone should enter modal state on touch.
    */
   enterModalSelectionOnTouch?: boolean;
+
+  /**
+   * Rerender DetailsRow only when props changed. Might cause regression when depending on external updates.
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts]: Would be good to document the default value via `@default`. This also makes it into the autodoc.

File: common/changes/office-ui-fabric-react/fix_detailslist_2018-05-13-19-01.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "office-ui-fabric-react",
+      "comment": "DetailsList: DetailsRow now behaves as a pure component, resulting in improved performance.",
+      "type": "patch"
```
Comment (cliffkoh) [inline common/changes/office-ui-fabric-react/fix_detailslist_2018-05-13-19-01.json]: Because of the new prop, no longer a patch change...

**Comment by aditima** [comment]
@OneCyrus - could you please sync and merge with master?

**Comment by dzearing** [comment]
@OneCyrus any update?

**Comment by OneCyrus** [comment]
i need to fix the merge for version 6. i'll have a look at the weekend.

**Comment by OneCyrus** [comment]
needed to create a new PR #5571. I'll close this one.
